internal/client/v1/client/filter/egress/client.go

```diff
@@ -0,0 +1,261 @@
+//
+// Copyright (C) 2019-2021 vdaas.org vald team <vald@vdaas.org>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Package egress provides egress filter client logic
+package egress
+
+import (
+	"context"
+	"reflect"
+	"strings"
+	"sync"
+
+	"github.com/vdaas/vald/apis/grpc/v1/filter/egress"
+	"github.com/vdaas/vald/apis/grpc/v1/payload"
+	"github.com/vdaas/vald/internal/errors"
+	"github.com/vdaas/vald/internal/net/grpc"
+	"github.com/vdaas/vald/internal/observability/trace"
+)
+
+type Client interface {
+	egress.FilterClient
+	Target(ctx context.Context, targets ...string) (egress.FilterClient, error)
+	GRPCClient() grpc.Client
+	Start(context.Context) (<-chan error, error)
+	Stop(context.Context) error
+}
+
+type client struct {
+	addrs []string
+	cl    sync.Map
+	c     grpc.Client
+}
+
+type specificAddrClient struct {
+	addr string
+	c    grpc.Client
+}
+
+type multipleAddrsClient struct {
+	addrs []string
+	c     grpc.Client
+}
+
+const (
+	apiName = "vald/internal/client/v1/client/filter/egress"
+)
+
+func New(opts ...Option) (Client, error) {
+	c := new(client)
+	for _, opt := range append(defaultOptions, opts...) {
+		err := opt(c)
+		if err != nil {
+			return nil, errors.ErrOptionFailed(err, reflect.ValueOf(opt))
+		}
+	}
+	if c.c == nil {
+		if c.addrs == nil {
+			return nil, errors.ErrGRPCTargetAddrNotFound
+		}
+		c.c = grpc.New(grpc.WithAddrs(c.addrs...))
+	}
+	return c, nil
+}
+
+func (c *client) Start(ctx context.Context) (<-chan error, error) {
+	return c.c.StartConnectionMonitor(ctx)
+}
+
+func (c *client) Stop(ctx context.Context) error {
+	return c.c.Close(ctx)
+}
+
+func (c *client) GRPCClient() grpc.Client {
+	return c.c
+}
+
+func (c *client) Target(ctx context.Context, targets ...string) (egress.FilterClient, error) {
+	if targets == nil {
```

Comment:
if I made a misstake, I am sorry:bow:
Should we check with length(len) ?
```go
len(targets) == 0
```
https://play.golang.org/p/wc04qE0JNLc